src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -220,13 +220,21 @@ private void CheckInvariants()
             solutionAttributes ??= _solutionAttributes;
             projectIds ??= ProjectIds;
             idToProjectStateMap ??= _projectIdToProjectStateMap;
-            options ??= Options.WithLanguages(GetRemoteSupportedProjectLanguages(idToProjectStateMap));
             analyzerReferences ??= AnalyzerReferences;
             projectIdToTrackerMap ??= _projectIdToTrackerMap;
             filePathToDocumentIdsMap ??= _filePathToDocumentIdsMap;
             dependencyGraph ??= _dependencyGraph;
             var newFrozenSourceGeneratedDocumentState = frozenSourceGeneratedDocument.HasValue ? frozenSourceGeneratedDocument.Value : _frozenSourceGeneratedDocumentState;
 
+            // PERF: Only invoke WithLanguages if we have different set of project IDs (AddProject/RemoveProject operation)
+            Debug.Assert(idToProjectStateMap.Count != _projectIdToProjectStateMap.Count ||
+                idToProjectStateMap.Keys.SetEquals(_projectIdToProjectStateMap.Keys));
+            Debug.Assert(idToProjectStateMap.Count != _projectIdToProjectStateMap.Count ||
+                Options == Options.WithLanguages(GetRemoteSupportedProjectLanguages(idToProjectStateMap)));
```

Comment:
Can we put these invariants in the CheckInvariants method? That way it'll get ran through places other than Branch, but actually at the constructor level.